arrows/mvg/algo/integrate_depth_maps.cxx

```diff
@@ -0,0 +1,336 @@
+void
+integrate_depth_maps::priv
+::integrate_depth_map(image_of<double>& volume,
+                      camera_perspective const& camera,
+                      image_of<double> const& depth,
+                      image_of<double> const& weight,
+                      vector_3d const& origin,
+                      vector_3d const& spacing) const
+{
+  auto const ni = volume.width();
+  auto const nj = volume.height();
+  auto const nk = static_cast<int64_t>(volume.depth());
+
+  auto const w = depth.width();
+  auto const h = depth.height();
+
+  #pragma omp parallel for
+  for (int64_t k = 0; k < nk; ++k)
+  {
+    double const z = origin[2] + (k + 0.5) * spacing[2];
+    for (size_t j = 0; j < nj; ++j)
+    {
+      double const y = origin[1] + (j + 0.5) * spacing[1];
+      for (size_t i = 0; i < ni; ++i)
+      {
+        double const x = origin[0] + (i + 0.5) * spacing[0];
+        vector_3d world_pt = vector_3d{ x, y, z };
+        vector_2d image_pt = camera.project(world_pt);
+        int const u = static_cast<int>(std::round(image_pt.x()));
+        int const v = static_cast<int>(std::round(image_pt.y()));
+        vector_2i pixel = image_pt.cast<int>();
+        if (u < 0 || u >= w ||
+            v < 0 || v >= h)
+        {
+          continue;
+        }
+        double const& d = depth(u, v);
+        double const a = (weight.size() > 0) ? weight(u, v) : 0.0;
+
+        if (d <= 0.0 || a <= 0.0)
+        {
+          continue;
+        }
+        double real_d = camera.depth(world_pt);
+        volume(i, j, k) += a * ray_potential(d, real_d);
+      }
+    }
+  }
+}
+
+// ----------------------------------------------------------------------------
+
+// compute the TSDF ray potential given an estimated depth and real depth
+double
+integrate_depth_maps::priv
+::ray_potential(double est_depth, double real_depth) const
+{
+  double diff = real_depth - est_depth;
+
+  double abs_diff = std::abs(diff);
+
+  if (abs_diff > const_delta)
+  {
+    return diff > 0.0 ? const_occluded_val
+                      : const_freespace_val;
+  }
+  else if (abs_diff > const_thickness)
+  {
+    return std::copysign(ray_potential_rho, diff);
+  }
+
+  return const_slope * diff;
+}
+
+// ----------------------------------------------------------------------------
+
+/// Constructor
+integrate_depth_maps::integrate_depth_maps()
+  : d_(new priv)
+{
+}
+
+// ----------------------------------------------------------------------------
+
+/// Destructor
+integrate_depth_maps::~integrate_depth_maps()
+{
+}
+
+// ----------------------------------------------------------------------------
+
+/// Get this algorithm's \link vital::config_block configuration block \endlink
+vital::config_block_sptr
+integrate_depth_maps::get_configuration() const
+{
+  // get base config from base class
+  auto config = vital::algo::integrate_depth_maps::get_configuration();
+
+  config->set_value("ray_potential_thickness", d_->ray_potential_thickness,
```

Comment:
These config values are all shared with the CUDA version.  It may make sense to make both of these classes share a common base class that handles common configuration.

Reply:
this sounds reasonable to me as well